docs/website/contents/for-developers/HardWonWisdom.md

```diff
@@ -348,3 +348,77 @@ The key corresponding parts of the implementation are as follows.
+      This use of `G`'s ticking rules instead of `F`'s to cross the end of the last epoch of `F` is subtle and somewhat counterintuitive (eg see Issue <https://github.com/input-output-hk/cardano-ledger/issues/3491>).
+    - A function `LedgerConfig F -> LedgerConfig G -> EpochNo {- start of G -} -> LedgerState F -> LedgerState G`.
+      This function will be applied to the ledger state immediately after the last `F` block.
+      Then the ledger rules for `G` will be used to tick that state to the slot of the first `G` block.
+      This use of `G`'s ticking rules instead of `F`'s to cross the end of the last epoch of `F` is subtle and somewhat counterintuitive (eg see Issue <https://github.com/input-output-hk/cardano-ledger/issues/3491>).
+    - Instances of `SingleEraBlock` for `F` and `G`, which decomposes as follows.
+        - Types that represent the ledger and protocol configuration _without_ any `EpochInfo` (see `HasPartialLedgerConfig` and `HasPartialConsensusConfig`).
+        - A function `singleEraTransition :: PartialLedgerConfig blk -> EraParams -> Bound {- era's start -} -> LedgerState blk -> Maybe EpochNo` which must return `Just e` if and only if the arguments determine that only a roll back of more than `k` blocks (which is impossible without disk corruption or manual intervention) could prevent this era from ending between epoch `e` and its predecessor (ie epoch `e-1` when `0<e` and "genesis" otherwise).
+            - This function must respect the safe zone in the given `EraParams`; see above.
+            - The logic used in this function must be invariant with respect to ticking.
+              In other words, if the end of the era is unknown or uncertain after the application of a block, then it cannot become certainly known before the next block.
+                - This constraint justifies that the function's type signature does not support applying it to a ticked ledger state; if it hypothetically did, then its output would exactly agree with the pre-image of that ticked ledger state.
+                  (Implementation detail: a ticked HFC ledger state must include `TransitionInfo` derived from the non-ticked ledger state only because the signature of `singleEraTransition` would not otherwise allow a ticked HFC ledger state's contents to determine the `TransitionInfo`.)
+                - Notably, this constraint is trivially true if the function is actually counting blocks in order to achieve its montonicity with respect to maximum rollbacks without relying on a Chain Growth property.
+
+See also "Which slots/times/epochs can we translate?" above.
+
+Some notable components are not listed above because they are determined by existing data about `F` and `G`.
+
+- The `TopLevelConfig (HardForkBlock [F, G])` can be completely determined from the same for `F` and `G`.
+- The initial state of the ledger and protocol can be completely determined from the same for `F`.
```

Comment:
Exception here is the skip-ahead benchmarking functionality, which relies on `singleEraTransition` for `F`.

Reply:
That's not actually contradictory to what I wrote. I wrote "can". Sneaky.

Reply:
(And I do discuss the skipping later.)